### ISSUE TITLE
routeProgress holding zero-values for distances since update from 19 to 22

### ISSUE THREAD

**Issue by Ohemgi** [mapbox/mapbox-navigation-android#1469]
![route_progress](https://user-images.githubusercontent.com/12342268/47553905-77f04180-d908-11e8-98e3-c409425c7aa4.png)

**Android API: compileVersion 28**
**Mapbox Navigation SDK version: 0.22.0**

### Summary / motivation
First things first:
- only using plain `MapboxNavigation` instance
- not using navigation view 
- the behavior (i gonna describe next) appears to happen after a direct update from `0.19.0` to `0.22.0`
- `0.19.0` worked just fine
- update was "required" due to the need for a better location projection / prediction
- no errors / exceptions on initializing, starting, handling navigation

Hey there :wave: ,

lately i had a talk with @danesfeder about the need for a kalman filter for smooth navigation transition and accurate location projection [in this issue](https://github.com/mapbox/mapbox-navigation-android/issues/54#issuecomment-432694559). We came to the conclusion that it would be best to update the navigation sdk, since the version `0.22.0` has major updates regarding these requirements.

The update was easy, everything compiles fine, the route is correctly received and drawn etc.pp. Calling `mapboxNavigation.startNavigation(directionsRoute)` runs like a charm and in my `SomeFragment implements ProgressChangeListener` i receive the same `onProgressChangeEvent` like before. But one major thing is different in this setup: the `RouteProgress` instance holds all route related values (legs, steps, upcoming steps etc.) but it seems like all the "calculated" distances, like `routeProgress.distanceRemaining` are 0.0 . I started to track that down to the `RouteProcessorRunnable` that was introduced after `0.19.0` and it seems like the status 

```
NavigationStatus status = mapboxNavigator.retrieveStatus(new Date(),options.navigationLocationEngineIntervalLagInMilliseconds());
RouteProgress routeProgress = routeProcessor.buildNewRouteProgress(status, route);
```

here is always zero.

I'm kinda lost, why the update results in that kind of "partial error" (especially that everything seems to work fine except the progress distance calculation).

The navigation startup call looks like that

```
MapboxNavigationOptions options = MapboxNavigationOptions.builder()
                    .defaultMilestonesEnabled(false)
                    .snapToRoute(true)
                    .build();
mMapboxNavigation = new MapboxNavigation(
                    getContext(),
                    getResources().getString(R.string.mapbox_access_token),
                    options
);
mMapboxNavigation.addMilestoneEventListener(mNavigationInformation);
mMapboxNavigation.addProgressChangeListener(mNavigationInformation);
mMapboxNavigation.addOffRouteListener(mNavigationInformation);
mMapboxNavigation.setLocationEngine(mLocationEngine);
mMapboxNavigation.startNavigation(directionsRoute);
```

br, Ben

**Comment by Ohemgi** [comment]
@Guardiola31337 @zugaldia as discussed i added it as an issue. Even though i tried to highlight all related information, feel free to ask me for further hints about this :smile:

**Comment by siavashabdoli** [comment]
Seems we have same problem and as long as we use NavigationView, user see default value in summaryBottomSheet (1min and 50meters) but after user start moving we get correct value for RouteProgress and NavigationView set correct value on SummaryBottomSheet as well.

**Comment by danesfeder** [comment]
Hey @Ohemgi @siavashabdoli we will continue to dig into this on our end - this issue does still seem to persist when navigation starts but we have not started tracking the user (0 values provided until tracking).  Thanks for this report!

**Comment by danesfeder** [comment]
@Ohemgi when you are calling `mMapboxNavigation.setLocationEngine(mLocationEngine);` do you know if the `LocationEngine` is pushing location updates at this point?  

@kevinkreiser does `Navigator` still need at least one location update in order to push the full route values?  Or should it push the full values regardless? 

I just did some quick testing with the `NavigationView` (`0.22.0`) and was having trouble reproducing.  We will continue to dig on this as it seems you are both able to still do so.

![screen shot 2018-10-29 at 10 23 19 am](https://user-images.githubusercontent.com/8434572/47656438-740d2b00-db65-11e8-8538-547aeefb47a6.png)

**Comment by kevinkreiser** [comment]
@danesfeder Nope it pushes the full route values as soon as you load a route with the `initialized` status. all statuses will stay `initialized` with only the location and bearing changing until you either go `offroute` (in which the values will stay full values) or you start `tracking` in which the values will be appropriate to your distance/time along the loaded route.

**Comment by kevinkreiser** [comment]
perhaps we should check the senario in which you initialize the route and then you ask for a status in the future. this should returned the same `initialized` status but we should check that!

**Comment by danesfeder** [comment]
@Ohemgi @siavashabdoli can either of you confirm this is still an issue?

**Comment by stanfordmark** [comment]
Hello,

I'm getting this issue.  In my project I am getting routeprogress objects with 0 for distance and 0 for duration.  

<img width="1134" alt="screen shot 2018-11-15 at 12 45 09 pm" src="https://user-images.githubusercontent.com/41921726/48580627-94aee200-e8d4-11e8-83ac-ba908ba631a9.png">
<img width="443" alt="screen shot 2018-11-15 at 12 45 18 pm" src="https://user-images.githubusercontent.com/41921726/48580629-94aee200-e8d4-11e8-8ee4-ec79c88d2795.png">
<img width="1400" alt="screen shot 2018-11-15 at 12 45 46 pm" src="https://user-images.githubusercontent.com/41921726/48580630-94aee200-e8d4-11e8-9bd7-c1e473f254c5.png">
<img width="974" alt="screen shot 2018-11-15 at 12 45 54 pm" src="https://user-images.githubusercontent.com/41921726/48580632-95477880-e8d4-11e8-9c53-72b9e5b589b8.png">
<img width="1382" alt="screen shot 2018-11-15 at 12 46 10 pm" src="https://user-images.githubusercontent.com/41921726/48580633-95477880-e8d4-11e8-9b92-79cd2edc8e0d.png">
<img width="625" alt="screen shot 2018-11-15 at 12 46 19 pm" src="https://user-images.githubusercontent.com/41921726/48580634-95477880-e8d4-11e8-81ae-4e0fe2ef2c1e.png">

**Comment by Ohemgi** [comment]
Hey @danesfeder ! Sorry just came back from a trip to japan :) I can confirm @stanfordmark : using version 22 im getting a routeprogress object holding only zero values, even if i have "progress along the route"

**Comment by Ohemgi** [comment]
@danesfeder i just missed your comment so as an addition: The same location engine is used for the location layer and the location updates seem to be fine code-wise and the location moves as it should.

**Comment by danesfeder** [comment]
Hey @Ohemgi @stanfordmark 👋 can both you try with `0.23.0`?  We added a `RouteProgress#currentState` that will give more insight into where we are as far as confidence along the route:

`0.23.0`:

<img width="953" alt="screen shot 2018-11-16 at 4 28 13 pm" src="https://user-images.githubusercontent.com/8434572/48648357-095d4b80-e9bd-11e8-9a83-a116b72ae104.png">

If ya'll don't mind posting that information (when you are seeing the 0 values), that would help us greatly here.  Thanks for the help / patience as we try to reproduce!

**Comment by Ohemgi** [comment]
Hey @danesfeder :wave: 
As you suggested i updated to Version `23` and checked the route state. It seems like it is flaged as invalid (see Screenshot). The route is one i used for several months as testroute so it did not change at all. Do you have a good hint where i can track down the reason for that?
![route_invalid](https://user-images.githubusercontent.com/12342268/48676017-f97e6c80-eb60-11e8-92d6-e0c551af8d77.png)

**Comment by danesfeder** [comment]
Hey @Ohemgi thanks for checking on this - it may just be a matter of updating your test fixture with a fresh Directions API response. 

@kevinkreiser do you mind outlining some possible scenarios where we will say the `DirectionsRoute` passed is not valid?  May gives us some clues as to why @Ohemgi's test route is invalid in `Navigator`'s eyes.  Thanks!

**Comment by kevinkreiser** [comment]
@Ohemgi you beat me to it! I wanted to mention that the `NavigationStatus` object has an attribute called `stateMessage`. When you send it json and it comes back with an `INVALID` state the message should have some text in there telling you something about what was wrong. Glad you found it!

**Comment by stanfordmark** [comment]
I'm not sure if that's the same issue I'm coming across then.  In my screen shots you can see I'm using navroute to generate the route and the 2 points for the navroute are valid points.

**Comment by Ohemgi** [comment]
Have you checked the state like suggested by @danesfeder and @kevinkreiser ? For me it was a very good starting point to track down the issue (regardless my own solution / fix :smile: )

**Comment by stanfordmark** [comment]
Sure, sorry was on vacation.  I'll take a look and get back to you.

**Comment by danesfeder** [comment]
@stanfordmark can you share any snippet for your implementation outside of https://github.com/mapbox/mapbox-navigation-android/issues/1469#issuecomment-439184427 to help us debug?

Are you using `NavigationView` or `MapboxNavigation` on it's own?

**Comment by stanfordmark** [comment]
Sure, I'm using the MapboxNavigation with the MapBox MapView, LocationComponent and NavigationMapRoute.

Here is how we set up the location engine and navigation:
<img width="940" alt="screen shot 2018-11-26 at 2 57 10 pm" src="https://user-images.githubusercontent.com/41921726/49047250-ad987c80-f18b-11e8-80c7-d214f023fc40.png">


Here is how we handle the mapMatching response:
<img width="1087" alt="screen shot 2018-11-26 at 2 53 11 pm" src="https://user-images.githubusercontent.com/41921726/49047037-0fa4b200-f18b-11e8-9568-deebd3041bc4.png">


Here is our onProgress change in the same Navigation Manager class:
<img width="1510" alt="screen shot 2018-11-26 at 2 52 03 pm" src="https://user-images.githubusercontent.com/41921726/49047105-3ebb2380-f18b-11e8-9d9b-67721a689a3b.png">

Here is how we set up the mapview
<img width="719" alt="screen shot 2018-11-26 at 2 51 18 pm" src="https://user-images.githubusercontent.com/41921726/49047132-54304d80-f18b-11e8-9d40-d93407c6dc26.png">
<img width="1205" alt="screen shot 2018-11-26 at 2 51 42 pm" src="https://user-images.githubusercontent.com/41921726/49047133-54304d80-f18b-11e8-99bf-230dbde7d286.png">
<img width="1510" alt="screen shot 2018-11-26 at 2 52 03 pm" src="https://user-images.githubusercontent.com/41921726/49047134-54304d80-f18b-11e8-9f58-0c90f192a686.png">


We are using the progress update to set our UI state.  Is there any other information or code that would help?  The issue is that in previous libraries this code had been working as we intended.  We recently updated the libraries to 0.21.0 to fix an index out of bounds error with the onProgress call and that seems to have the side effect of giving us zeroes for distance and duration.  When upgrading to 0.23.0 we don't seem to get progress updates.

**Comment by Ohemgi** [comment]
@stanfordmark can you add `mainNavigation.addProgressChangeListener(this)` and try again?

**Comment by stanfordmark** [comment]
We are setting the addProgressChangeListener in the second image I attached in the previous comment. When we get the route response and null check it we then set the listeners and start navigation. Should this be done somewhere else or is there something I'm missing?

**Comment by Ohemgi** [comment]
Nah, i don't think so. I just missed your second init part. I can verify that aslong there is a pushing LocationEngine and a MapboxNavigation you get RouteProgress updates immediately.

**Comment by stanfordmark** [comment]
Thanks for the help, I know this is outside the scope of the original issue.  When using 6.5.0 and 0.22.0 I am still seeing the zero values.  When upgrading to 6.6.1 and 0.23.0 I am not getting location updates or progress updates.  Is there anything different between the 2 libraries that I may be overlooking?

**Comment by danesfeder** [comment]
@stanfordmark I think this has to do with the setup of the `LocationEngine` in the first screenshot. 
You should be calling `LocationEngine#activate` after a `LocationEngineListener` has been added.  And then in `LocationEngineListener#onConnected` you can call `LocationEngine#requestLocationUpdates` (essentially after your connection with Google API services has been established).

**Comment by stanfordmark** [comment]
Thanks for help and for getting back so quickly.  I have moved the active to after the setting of the listener.  I still am not getting updates.  I stepped through the flow and I can confirm that we get an onConnected callback and call `        locationEngine?.requestLocationUpdates()
` I will try on different devices to see if it may be an Android/Google Play Services related issue.  Are there any differences between 0.22.0 and 0.23.0 that I may be overlooking? The only API update I saw was in the RouteUtils().isArrivalEvent(routeProgress) no longer needing a milestone.

**Comment by stanfordmark** [comment]
Ok, I have an interesting update to report.  On an Asus P00I 10" running Android 7, API 24, the new location engine doesn't seem to work and the progress updates do not get called.  I used a mock location engine to test this.  When using the emulator, with the same mock location engine, I do see the progress updates and the progress state.  It seems to work as intended on a 10" emulator running API 27.  It does not run as expected using an emulator on API 28.

**Comment by danesfeder** [comment]
@stanfordmark we have observed some behaviors similar to what you're describing with our `core-location` library.  At times, there's a ~30 second delay before `Location` updates begin streaming correctly.  Our team that works on this library has already done some refactoring to address this, but the Navigation SDK is blocked at the moment while these changes land in the Maps SDK upstream first.  

In the meantime, you can try out our [`FusedLocationEngine`](https://github.com/mapbox/mapbox-navigation-android/blob/master/app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/location/FusedLocationEngine.java) in the test app.  This engine is compatible with the latest fused location APIs from play services.  We have seen much better start up GPS from this engine (no delay after the first update).  Thanks for your patience! 

cc @andrlee @electrostat

**Comment by stanfordmark** [comment]
@danesfeder This looks very promising.  I just copy pasta'd the code from the sample app, converted it to kotlin, grabbed the new engine and it immediately worked.  So far so good.  I'll write an update shortly as I step through the debugger and run some test routes in my mock location apparatus.

**Comment by stanfordmark** [comment]
So far so good.    It seems to be working quite well, the location is updating on the interval I set and I get good progress updates.  I have a question, there aren't options for the offroute detection anymore in the nav option builder.  Is there a way to change how long it takes or the threshold for offroute?  Besides that small issue with offroute, it seems that this ticket has a viable workaround.  Thanks @danesfeder

**Comment by danesfeder** [comment]
@stanfordmark great glad to hear that's working better now.  We recently did a major overhaul over the client-side routing logic to use a smarter, much more performant native library.  It's also now shared between iOS and Android for platform consistency.  When we did this, we closed some of those configuration params (like the off-route detection one you're referring to).  We are actively testing and watching how the new code performs - with some time, we are looking to expose certain variables / thresholds again.  I hope you understand and thank you for your patience.  

I'm going to close here now as resolved.  Thanks for the feedback and working through these issues @stanfordmark @Ohemgi
